### ISSUE TITLE
Ignore url query parameters when storing models (.mar files)

### ISSUE THREAD

**Issue by MuhsinFatih** [pytorch/serve#2085]
### 🚀 The feature

Ignore url query parameters when registering/accessing/deleting models. Instead of saving `.mar` files as `filename.mar&queryparams`, save them as `filename.mar`

### Motivation, pitch

I'm using S3 presigned URL's to access models stored on S3. I'm aware of the [Encrypted model serving](https://pytorch.org/serve/management_api.html#encrypted-model-serving) option which allows accessing encrypted models on S3, however this method requires that I allow torchserve access to the bucket containing my models. This is not useful for my use case because the software architecture I am working with uses pre-signed AWS url's supplied to torchserve by an API that governs access to models on S3. This is to follow the least privilege principle, thus torchserve instance doesn't have access to the production bucket except with presigned and short-lived url's.

The problem with using S3 presigned URL's is that torchserve registers models using a filename that includes the entirety of `filename.mar&queryparams`. Since query parameters change every time I want to access the model, the model-store cache always misses.

Example output of `ls model-store`:
```bash
╰─ ls model-store 
'model1.mar?AWSAccessKeyId=access-key&Signature=signature&Expires=expdate'
'model1.mar?AWSAccessKeyId=access-key&Signature=signature2&Expires=expdate2'
'model1.mar?AWSAccessKeyId=access-key&Signature=signature3&Expires=expdate3'
...
```

In simpler terms, this is what happens step-by-step:
1. My API generates presigned S3 url: `https://bucketname.s3.amazonaws.com/filename.mar?AWSAccessKeyId=****&Signature=****&Expires=****`
2. Torchserve receives url
3. Torchserve downloads the model. I expect it to save it as `filename.mar`, but torchserve saves it as `filename.mar?AWSAccessKeyId=****&Signature=****&Expires=****`
4. I do this again with another model. I then want to use the first model again. I expect a cache hit (I expect that torch-serve finds the model in model-store and avoids redownloading)
5. Torchserve receives a new presigned url for the same file. It's filename is again `filename.mar?AWSAccessKeyId=****&Signature=****&Expires=****`, but the signature and expiration fields are different, thus cache misses
6. Torchserve downloads the model again. Cache is cluttered because we're downloading the same model over and over again every time this happens, and we're waiting long download times

Proposed solution:
Ignore the query parameters in the url when saving the model. The filename is used [here](https://github.com/pytorch/serve/blob/2d1bd2649583c5a505840dcb1fbf9f812ced08fd/frontend/archive/src/main/java/org/pytorch/serve/archive/model/ModelArchive.java#L55) when registering a model and [here](https://github.com/pytorch/serve/blob/2d1bd2649583c5a505840dcb1fbf9f812ced08fd/frontend/archive/src/main/java/org/pytorch/serve/archive/model/ModelArchive.java#L147) when deleting a model. It's using [this](https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FilenameUtils.html#getName(java.lang.String)) method to get filename from url, but that's a function that's designed to read filenames from OS paths, not urls. Using something like [this](https://stackoverflow.com/a/17167743/2770195) should fix the filename

I can create a PR for this, but this is the first time I'm looking at the torchserve source code and I'd like to know; 1: If this solution is reasonable, 2: If I should consider other parts of the code that might need to change, 3: If anyone can suggest an alternative that will render this solution unnecessary

Thank you!

### Alternatives

Alternative solutions could be as follows:
### Use content-disposition-filename Header
I attempted to use the [content-disposition-filename](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#:~:text=explicit%20charset%20information.-,filename,-Is%20followed%20by) header when supplying S3 presigned url, but torchserve ignored the header. It might be better to check for the content-disposition header, and if it exists, use that to store the model

### Use model_name to store model file
Instead of scraping the model name from the url, use the provided model_name in the [model registration endpoint (`POST /models`)](https://pytorch.org/serve/management_api.html#register-a-model).

### Take filename as function argument when registering a model
Take optional filename argument in the [model registration endpoint (`POST /models`)](https://pytorch.org/serve/management_api.html#register-a-model). If it is provided, it would use that for model filename.

In my opinion this alternative is better than changing the existing filename convention because this option wouldn't be a breaking change

**Comment by mreso** [comment]
Hi,

thanks for filing the issue! The behavior is indeed unfortunate in your case and the proposed solution sounds reasonable to me.
The only edge case I could come up with is if somebody uses bogus parameters to create new filenames for model with the same name when he can not control the filenames on the server. But thats very exotic.

**Comment by MuhsinFatih** [comment]
@mreso awesome! I'll try to make a PR when I get the chance. I believe the bogus filename problem is best solved with the alternative number 3 (Take optional filename argument in the [model registration endpoint](https://pytorch.org/serve/management_api.html#register-a-model)). It would be the most flexible option for the developer as well. What do you think?

**Comment by mreso** [comment]
@MuhsinFatih Great! Looking forward to your PR! Let me know if you need any help with it!

Agreed on the solution for users using bogus filenames. Would be great to have this option in case users leverage that "flaw" in TS to make it work in their situation. If you want to take it on as well I would suggest creating a second PR for it to facilitate the review of the changes.

**Comment by cfculhane** [comment]
Hi @MuhsinFatih - curious how you got presigned S3 URL's to work in the first place? I keep hitting the same errors as https://github.com/pytorch/serve/issues/1293 , and even if the cache was missed this would allow us to move forward with serving models from S3..

**Comment by MuhsinFatih** [comment]
@cfculhane what's your S3 region? You might be having trouble with the signature version and/or with url parsing. Try the two following:
```
# specify signature version
s3_client = boto3.client("s3", config=Config(signature_version="s3v4"), region_name="your-region")
```

```
# replace & with %26
model_url = model_url.replace("&", "%26")
```
(source: https://github.com/pytorch/serve/issues/669#issuecomment-688995028)

**Comment by cfculhane** [comment]
Ahh thank you - this worked.  I ended up coding up a fastAPI web service inside the torchserve container to handle these and download them to the model store, which allowed us to cache them better and to handle both the preseigned urls without modification, and also S3 URI's natively. We then send the register model POST request off with just the local filename, and torchserve is none the wiser :)
